Add support for tflint-ignore-file annotations in JSON (#2230)
terraform-linters/tflint: docs/user-guide/annotations.md

  instance_type = "t1.2xlarge"
}
```

The `tflint-ignore-file` annotation is also supported in Terraform JSON by
using a top-level [comment property](https://developer.hashicorp.com/terraform/language/syntax/json#comment-properties):

```json
{
  "//": {

>     "tflint-ignore-file": "aws_instance_invalid_type"

Reviewer: Are there any other examples you can offer of this map comment structure being used? The documentation illustrates comment properties with string values. In that sense, this proposal clashes with the official guidance, as a JSON document that already contains a comment property (string) would be unable to conform to this object syntax.

Seems like `//` could be parsed equivalently to an HCL document with a first-line comment.

Author: I have no examples of this being used "in the wild"; my reasoning for the map structure was only that, hypothetically, someone might want to be able to use the comment property to leave other information. It wasn't clear to me how that would work if the comment property was a string which was already occupied by the `tflint-ignore-file` annotation.

For example:

```json
{
  "//": {
    "tflint-ignore-file": "foo",
    "_1": "we ignore foo because reasons",
    "_2": "this file does stuff"
  }
}
```

I'm realizing now though that the JSON spec may allow for multiple properties with the key `//`; that may be a cleaner way to have multiple comments, at the expense of being harder to parse here (`encoding/json` discards duplicate entries). I could look for a way of handling that though. E.g.

```json
{
  "//": "this file does stuff",
  "//": "we ignore foo because reasons",
  "//": "tflint-ignore: foo"
}
```

I'm open to alternatives though, it might make sense to expect a escaped linefeed-delimited string or a list of strings. E.g.

```json
{
  "//": "this file does stuff\nwe ignore foo because reasons\ntflint-ignore: foo"
}
```

or

```json
{
  "//": [
    "this file does stuff",
    "we ignore foo because reasons",
    "tflint-ignore: foo"
  ]
}
```

Reviewer: This was my assumption, thanks. I'd treat it as a string then, with the same rules as HCL:

https://github.com/terraform-linters/tflint/blob/fcbcce59c42d22c8ea69697c3c81ae85ee941413/tflint/annotation.go#L49

So if you have some existing comment property in the config, `tflint-ignore-file` would have to be prepended to that. Followed by whitespace, then any other comment content.

The map or list patterns are not inherently unreasonable, but TFLint tries to avoid creating Terraform patterns of its own.

Author: Makes sense to me; will update in a bit here.

Reviewer 2: > The documentation illustrates comment properties with string values

I thought the same thing, but as mentioned in https://github.com/terraform-linters/tflint/pull/2230#discussion_r1949900509, I thought a map would not be a bad choice to add other comments.

In fact, having looked at the HCL JSON parser, it doesn't seem to matter what the type of the comment value is:
https://github.com/hashicorp/hcl/blob/v2.23.0/json/structure.go#L55-L59

Personally, I'm concerned that comments like the following look like comments to the `tflint-ignore-file` annotation:

```json
{
  "//": "tflint-ignore-file: aws_instance_invalid_type # This file is auto-generated by tools/terraform_gen.go. DO NOT EDIT",
}
```

I like the following style the most:

```json
{
  "//": "tflint-ignore-file: aws_instance_invalid_type",
  "//": "This file is auto-generated by tools/terraform_gen.go. DO NOT EDIT"
}
```

However, given the difficulties involved when you have multiple keys with the same name, I believe we need to find some compromise.

Reviewer: All these options have sharp edges. Ideally I think you'd have this config if you wanted an ignore and an unrelated comment:

```json
{
  "//": "tflint-ignore-file: aws_instance_invalid_type\nThis file is auto-generated by tools/terraform_gen.go. DO NOT EDIT",
}
```

If you put them on the same line it would imply that the comment described the ignore.

If you have some existing code that generates a comment property, you can just update it to `ignore + "\n" + existing` and you're set. Expecting a map or array requires more refactoring, even if it's ultimately very minor.

Ultimately JSON is meant for generated config and not humans anyway and so I err on the side of "good enough" versus something more expressive/structured.

  },
  "resource": {
    "aws_instance": {
      "foo": {
        "instance_type": "t2.micro"
      }
    }
  }
}
```

> Note that the `tflint-ignore` annotation is not currently supported in
> Terraform JSON config.

Reviewer: ```suggestion
The `tflint-ignore` annotation is not supported JSON configuration.
```

Avoid "currently" in documentation and similar caveats. Just say how it works right now.
